src/lib/utils/time.ts

```diff
@@ -0,0 +1,89 @@
+export const millisecondsPerSecond = 1000
+export const secondsPerMinute = 60
+export const minutesPerHour = 60
+export const hoursPerDay = 24
+export const daysPerWeek = 7
+
+export const millisecondsPerMinute = millisecondsPerSecond * secondsPerMinute
+export const millisecondsPerHour = millisecondsPerMinute * minutesPerHour
+export const millisecondsPerDay = millisecondsPerHour * hoursPerDay
+export const millisecondsPerWeek = millisecondsPerDay * daysPerWeek
+export const previousSevenDays = millisecondsPerWeek
+export const lastThirtyDays = millisecondsPerDay * 30
+
+export const currentYear = new Date().getFullYear()
+
+export const millisecondsSince = (date: Date) => {
+  const now = new Date()
+  return date > now ? 0 : now.getTime() - date.getTime()
+}
+
+export const timeSince = (date: Date) => {
+  const timeDifference = millisecondsSince(date)
+  while (timeDifference != 0) {
+    if (timeDifference < millisecondsPerMinute) {
+      return `${Math.floor(timeDifference / millisecondsPerSecond)}s`
+    } else if (timeDifference < millisecondsPerHour) {
+      return `${Math.floor(timeDifference / millisecondsPerMinute)}m`
+    } else if (timeDifference < millisecondsPerDay) {
+      return `${Math.floor(timeDifference / millisecondsPerHour)}h`
+    } else if (timeDifference < millisecondsPerWeek) {
+      return `${Math.floor(timeDifference / millisecondsPerDay)}d`
+    } else {
+      return `${Math.floor(timeDifference / millisecondsPerWeek)}w`
+    }
+  }
+  return 0
+}
+
+export const categorizeDate = (date: Date) => {
+  if (isToday(date)) {
+    return 'today'
+  } else if (isYesterday(date)) {
+    return 'yesterday'
+  } else if (
+    millisecondsPerDay < millisecondsSince(date) &&
+    millisecondsSince(date) <= previousSevenDays
+  ) {
+    return 'previousSevenDays'
+  } else if (
+    previousSevenDays < millisecondsSince(date) &&
+    millisecondsSince(date) <= lastThirtyDays
+  ) {
+    return 'lastMonth'
+  } else if (date.getFullYear() >= currentYear) {
+    const year = date.getFullYear()
+    return year
+  } else {
+    return 'undefined'
```

Comment:
Returning "undefined" as a string is a bit confusing, as `undefined` (no quotes) is a value as well. This case means the date is in the future, right? Then I would return either empty string, or possibly "inTheFuture" or something.

Reply:
i added a check to see if the date provided was actually in the future because i don't think the current way explicitly caught future dates